memtable/write_buffer_manager.cc

```diff
@@ -64,127 +97,96 @@ void WriteBufferManager::ReserveMem(size_t mem) {
   if (enabled()) {
     memory_active_.fetch_sub(mem, std::memory_order_relaxed);
   }
 }
 
 void WriteBufferManager::FreeMem(size_t mem) {
+  std::unique_lock<std::mutex> lock(wbm_mutex_);
   if (cache_res_mgr_ != nullptr) {
     FreeMemWithCache(mem);
   } else if (enabled()) {
     memory_used_.fetch_sub(mem, std::memory_order_relaxed);
   }
-  // Check if stall is active and can be ended.
   MaybeEndWriteStall();
 }
 
-void WriteBufferManager::FreeMemWithCache(size_t mem) {
-  assert(cache_res_mgr_ != nullptr);
-  // Use a mutex to protect various data structures. Can be optimized to a
-  // lock-free solution if it ends up with a performance bottleneck.
-  std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
-  size_t new_mem_used = memory_used_.load(std::memory_order_relaxed) - mem;
-  memory_used_.store(new_mem_used, std::memory_order_relaxed);
-  Status s = cache_res_mgr_->UpdateCacheReservation(new_mem_used);
-
-  // We absorb the error since WriteBufferManager is not able to handle
-  // this failure properly.
-  // [TODO] We'll need to improve it in the future and figure out what to do on
-  // error
-  s.PermitUncheckedError();
+bool WriteBufferManager::MaybeBeginWriteStall(StallInterface* wbm_stall) {
+  assert(wbm_stall != nullptr);
+  std::unique_lock<std::mutex> lock(wbm_mutex_);
+  if (ShouldStall()) {
+    wbm_stall->SetState(StallInterface::State::BLOCKED);
+    std::list<StallInterface*> new_node = {wbm_stall};
+    queue_.splice(queue_.end(), std::move(new_node));
+    return true;
+  } else {
+    return false;
+  }
 }
 
-void WriteBufferManager::BeginWriteStall(StallInterface* wbm_stall) {
+void WriteBufferManager::RemoveDBFromQueue(StallInterface* wbm_stall) {
   assert(wbm_stall != nullptr);
-  assert(allow_stall_);
-
-  // Allocate outside of the lock.
-  std::list<StallInterface*> new_node = {wbm_stall};
 
+  // Deallocate the removed nodes outside of the lock.
+  std::list<StallInterface*> cleanup;
   {
-    std::unique_lock<std::mutex> lock(mu_);
-    // Verify if the stall conditions are stil active.
-    if (ShouldStall()) {
-      stall_active_.store(true, std::memory_order_relaxed);
-      queue_.splice(queue_.end(), std::move(new_node));
+    std::unique_lock<std::mutex> lock(wbm_mutex_);
+    if (enabled() && allow_stall_.load(std::memory_order_relaxed)) {
+      for (auto it = queue_.begin(); it != queue_.end();) {
+        auto next = std::next(it);
+        if (*it == wbm_stall) {
+          cleanup.splice(cleanup.end(), queue_, std::move(it));
+        }
+        it = next;
+      }
     }
   }
-
-  // If the node was not consumed, the stall has ended already and we can signal
-  // the caller.
-  if (!new_node.empty()) {
-    new_node.front()->Signal();
-  }
+  wbm_stall->Signal();
 }
 
-// Called when memory is freed in FreeMem or the buffer size has changed.
 void WriteBufferManager::MaybeEndWriteStall() {
-  // Cannot early-exit on !enabled() because SetBufferSize(0) needs to unblock
```

Comment:
@akankshamahajan15  - I am not quite clear about this comment. Does it means we don't need to early-exit on !enabled() because it won't happen as "SetBufferSize(0) needs to unblock...."? Or does it mean we is not able to do so because we don't meet the requirement of "SetBufferSize(0) needs to unblock..."?

Either way, I think SetBufferSize(0) can happen anytime as it can be called by users as long as they have access to the WBM ...